cli/src/main/java/com/devonfw/tools/ide/context/GitContext.java

```diff
@@ -0,0 +1,82 @@
+package com.devonfw.tools.ide.context;
+
+import java.nio.file.Path;
+
+import com.devonfw.tools.ide.log.IdeLogger;
+
+/**
+ * Interface for git commands with input and output of information for the user.
+ */
+public interface GitContext extends IdeLogger {
+
+  /**
+   * Checks if the Git repository in the specified target folder needs an update by inspecting the modification time of
+   * a magic file.
+   *
+   * @param repoUrl the git remote URL to clone from. May be suffixed with a hash-sign ('#') followed by the branch name
+   *        to check-out.
+   * @param targetRepository the {@link Path} to the target folder where the git repository should be cloned or pulled.
+   *        * It is not the parent directory where git will by default create a sub-folder by default on clone but the *
+   *        final folder that will contain the ".git" subfolder.
+   * @param remoteName the remote server name.
+   * @param branchName the name of the branch.
+   * @param force boolean true enforces a git hard reset and cleanup of added files.
+   */
+  void pullOrCloneIfNeeded(String repoUrl, Path targetRepository, String remoteName, String branchName, boolean force);
```

Comment:
we should discuss this.
* Do we really always need this in our signatures here?
* What if I want to clone the default branch of a repository but I do not know the name of that default branch? That currently does not really be possible.
* Looking at the implementation these 2 new parameters are however ignored unless `force` is `true`. That is completely unclear for the API user reading the JavaDoc.
* Should we remove the `force` flag and these 2 new parameters entirely and instead add a new method signature for the `pullOrFetchAndResetIfNeeded` usecase? Or is there a better idea?

Reply:
The `reset` method can stay as it is. A `pullOrFetchAndResetIfNeeded` may have an optional `remoteName` argument and use `"origin"` as default. Then all would be consistent and clean.

Reply:
Thanks for your suggestions. I've now implemented them, please check again.